### PR TITLE
GPU Attribute Improvements

### PR DESCRIPTION
**Pull request by DanilaFe**
Closes https://github.com/chapel-lang/chapel/issues/22822.

This PR seeks to improve the experience of using various GPU attributes in several cases. These changes are motivated by some of the difficulties I've observed @Iainmon to have had while working on his GPU-enabled machine learning code.

## Case 1: Writing CPU and GPU code using the GPU locale model
The way to ensure that a loop is GPU-eligible in user code (and to fail compilation if the loop is not GPU eligible), is to use `@assertOnGpu`. However, one cannot do this when writing code that is expected to support both GPUs and CPUs. I've observed Iain's code to have something like the following:

```Chapel
if onGpu {
  @assertOnGpu
  foreach ... { }
} else {
  foreach ... { /* the same loop as above */ }
}
```

This way, the code could be used on both the GPU and the CPU, and the compiler will ensure that the GPU version is eligible. However, this introduces a maintenance burden, and makes the code rather verbose. To work around this problem, I introduce a new GPU primitive + attribute: `@gpu.assertEligible`. This attribute has the same behavior as `@assertOnGpu` at compile-time, but it does not have a runtime effect. Thus, the code above can be flattened and continue to support both CPU and GPU runs:

```Chapel
if onGpu {
  @gpu.assertEligible
  foreach ... { }
}
```

In my opinion, we should phase out the use of `@assertOnGpu` in favor of `@gpu.assertEligible`. It's unclear to me that having a runtime assertion using this attribute is worth keeping it around two similar attributes. Personally, I think that the compile-time assertion can be handled by `@gpu.assertEligible,` and various utilities from `GpuDiagnostics` for tracking kernel launches etc. can be used to ensure that GPU execution occurs at runtime. This PR doesn't make this (potentially more controversial change).

## Case 2: Disabling GPU support and compiling with `CHPL_LOCALE_MODEL=flat`
When I told Iain to run his performance experiments in the flat locale model (to get started with initial performance results via the CPU), he immediately ran into internal errors. This is an instance of https://github.com/chapel-lang/chapel/issues/22822.

My chosen solution to this problem is to make `@assertOnGpu` a compile-time error under `CHPL_LOCALE_MODEL=flat`. This is because of the semantics of `@assertOnGpu`: this attribute has a runtime check; without a GPU, the check is guaranteed to fail, and cause a "certain" failure. This error is now user-facing, and tells the user to switch to `@gpu.assertEligible` if all they want is a compile-time check.

On the other hand, the `@gpu.assertEligible` attribute, which does not have any runtime semantics, does not cause a compilation error with the `flat` locale model. Instead, the attribute is simply ignored (we don't perform any GPU logic with the flat locale model, and it doesn't seem worth it to actually perform GPU transformations / analysis for the sole purpose of validating GPU eligibility). The same is true for `@gpu.blockSize`, and the non-user-facing "GPU primitive block" primitive which is used to group GPU primitives created via attributes. Thus, the following code compiles and runs just fine in the `flat` locale model:

```Chapel
@gpu.assertEligible
@gpu.blockSize(128)
foreach i in 1..128 { /* ... */ }
```

Reviewed by @e-kayrakli -- thanks!

## Testing
- [x] new `flat` tests for GPU primitives, including a new user-facing error.
- [x] GPU tests, including new tests for `@gpu.assertEligible`
- [x] paratest

### CODE REVIEW

File: compiler/optimizations/gpuTransforms.cpp [hunk 1]
```diff
@@ -2144,8 +2155,20 @@ static void cleanupPrimitives() {
...
       // the copy.
       cleanupTaskIndependentCapturePrimitive(callExpr);
-    }
-    else if(callExpr->isPrimitive(PRIM_GPU_SET_BLOCKSIZE)) {
+    } else if (callExpr->isPrimitive(PRIM_GPU_SET_BLOCKSIZE) ||
+               callExpr->isPrimitive(PRIM_ASSERT_GPU_ELIGIBLE)) {
+      callExpr->remove();
+    } else if (callExpr->isPrimitive(PRIM_ASSERT_ON_GPU)) {
+      if (!usingGpuLocaleModel()) {
+        USR_FATAL_CONT(callExpr, "@assertOnGpu encountered in non-GPU compilation");
```
Comment (e-kayrakli) [inline compiler/optimizations/gpuTransforms.cpp:2163]: I am still trying to decide whether this should be an error or warning. Technically, my `assertOnGpu`-tagged loop could be out of the dynamic execution path. That could appear in code like

```chpl
if here.gpus.size == 0 {
  foreach
}
else {
  @assertOnGpu
  foreach
}
```

I know that this is not the ideal way of performing that check, but it is perfectly valid nonetheless.

As a sidenote, Andy has this dream where the GPU locale model will be the only locale model, and you'll make decisions like the one above. We shouldn't make decisions based on that, but thought I should mention.

The bottom line to me is, if you have code like that today with the buggy `assertOnGpu`, it would work. I believe it only failed during execution time with the `flat` locale model, but I may be misremembering.

File: compiler/passes/convert-uast.cpp [hunk 2]
```diff
@@ -126,6 +126,8 @@ struct LoopAttributeInfo {
   LLVMMetadataList llvmMetadata;
   // The @assertOnGpu attribute, if one is provided by the user.
   const uast::Attribute* assertOnGpuAttr = nullptr;
+  // The @gpu.assertEligible attribute, if one is provided by the user.
+  const uast::Attribute* assertEligibleAttr = nullptr;
```
Comment (e-kayrakli) [inline compiler/passes/convert-uast.cpp:130]: Not a big worry, but this variable is not self-documenting at all. Eligible for what?

File: compiler/passes/convert-uast.cpp [hunk 3]
```diff
@@ -4366,12 +4367,16 @@ struct Converter {
 };
 
 bool LoopAttributeInfo::insertGpuEligibilityAssertion(BlockStmt* body) {
+  bool inserted = false;
   if (assertOnGpuAttr) {
-    body->insertAtTail(new CallExpr(PRIM_ASSERT_ON_GPU,
-                                    new SymExpr(gTrue)));
-    return true;
+    body->insertAtTail(new CallExpr("chpl__assertOnGpuAttr"));
```
Comment (e-kayrakli) [inline compiler/passes/convert-uast.cpp:4372]: I know I said this wouldn't bother me, but could you explain to me (or point me to an issue) why this is better?

File: compiler/passes/convert-uast.cpp [hunk 4]
```diff
@@ -4366,12 +4367,16 @@ struct Converter {
 };
 
 bool LoopAttributeInfo::insertGpuEligibilityAssertion(BlockStmt* body) {
+  bool inserted = false;
   if (assertOnGpuAttr) {
-    body->insertAtTail(new CallExpr(PRIM_ASSERT_ON_GPU,
```
Comment (e-kayrakli) [inline compiler/passes/convert-uast.cpp:4370]: Have you removed this primitive? I don't think you did.

File: compiler/passes/convert-uast.cpp [hunk 5]
```diff
@@ -4384,16 +4389,11 @@ bool LoopAttributeInfo::insertBlockSizeCall(Converter& converter, BlockStmt* bod
   static int counter = 0;
 
   if (blockSizeAttr) {
-    if (blockSizeAttr->numActuals() != 1) {
-      USR_FATAL(blockSizeAttr->id(),
-                "'@gpu.blockSize' attribute must have exactly one argument: "
-                "the block size");
```
Comment (e-kayrakli) [inline compiler/passes/convert-uast.cpp:4390]: Do we have this check elsewhere now? I may have missed it.

Comment (e-kayrakli) [reply compiler/passes/convert-uast.cpp:4390]: Do we have a test for it? If not, maybe we should (ok to add in this PR).

File: modules/standard/AutoGpu.chpl [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+pragma "module included by default"
+@unstable("The module name 'AutoGpu' is unstable.")
```
Comment (e-kayrakli) [inline modules/standard/AutoGpu.chpl:21]: As this is an auto-included module and we already have an unstable warning for the GPU locale model, I think this should be removed.

File: modules/standard/AutoGpu.chpl [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+module AutoGpu {
+  use ChplConfig;
+  use Errors;
+
```
Comment (e-kayrakli) [inline modules/standard/AutoGpu.chpl:25]: Could you add a comment here or at the module header explaining that calls to these functions are added by the compiler?

File: compiler/passes/convert-uast.cpp (hunk 5 above)
Comment (e-kayrakli) [reply compiler/passes/convert-uast.cpp:4390]: NVM. I see where you added the check, and can see that as a benefit of the function-based vs primitive-based approach. But, I'd still like to hear from you other benefits if any.

**Review by e-kayrakli — COMMENTED** [review]
Looks good generally. I have mostly nits. I have more testing ideas that I'd like to see go in with this PR, though:

- Error message for `blockSize(1,2)` if we don't have it. (I know it is not about this PR, but the PR touches relevant areas)
- More importantly, what happens if I use both flavors of assertions together in flat and gpu locale models? Some of those cases will error, some should just work. It'd be good to see that they can be used without an embarrassing issue.

File: modules/standard/AutoGpu.chpl (hunk 6 above)
Comment (DanilaFe) [reply modules/standard/AutoGpu.chpl:21]: I believe this warning only flags if you explicitly use `AutoGpu`; the same is true for `AutoMath`, which is where I got this from. Hence, the module _name_ is unstable.

File: compiler/optimizations/gpuTransforms.cpp [hunk 8]
```diff
@@ -410,11 +410,17 @@ class GpuAssertionReporter {
...
-    auto isAttributeSym = toSymExpr(assertion->get(1));
-    INT_ASSERT(isAttributeSym);
-    if (isAttributeSym->symbol() == gTrue) {
-      reason = "is marked with @assertOnGpu";
+    const char* reason = nullptr;
+    if (assertion->isPrimitive(PRIM_ASSERT_GPU_ELIGIBLE)) {
+      reason = "is marked with @gpu.assertEligible";
+    } else {
+      CHPL_ASSERT(assertion->isPrimitive(PRIM_ASSERT_ON_GPU));
+      reason = "contains assertOnGpu()";
```
[the remark this answers is not in the document]
Comment (DanilaFe) [reply compiler/optimizations/gpuTransforms.cpp:418]: We do; the legacy `assertOnGpu` is still present in the code. I deprecated and can remove it in a separate PR, but would rather not do it in this one.

File: compiler/passes/convert-uast.cpp (hunk 3 above)
Comment (DanilaFe) [reply compiler/passes/convert-uast.cpp:4372]: Some of my reasons:

1. I can use the overload resolution process to select candidate functions. For instance, I may use a where clause to avoid inserting GPU primitives when CHPL_LOCALE_MODEL=flat
2. I can ensure that the arguments to the primitives are correctly typed (since calls are just “regular functions”).
3. I don’t have to change compiler code to adjust the behavior of attributes, to some extent
4. Using an experimental annotation (which I plan to work on next after this PR) I can get `chpldoc` to use these function definitions (like `chpl__assertOnGpuAttr`) to generate documentation for attributes, which addresses https://github.com/chapel-lang/chapel/issues/23149.

File: compiler/passes/convert-uast.cpp (hunk 4 above)
Comment (DanilaFe) [reply compiler/passes/convert-uast.cpp:4370]: I didn't, and don't intend to in this PR, though my goal in the long term is to remove it.

File: compiler/optimizations/gpuTransforms.cpp (hunk 1 above)
Comment (DanilaFe) [reply compiler/optimizations/gpuTransforms.cpp]: To make it a runtime warning and allow the program to compile, I needed to add an implementation of `assert_on_gpu` for CPU specifically. I did so by tacking on to `chpl-gpu.h`, but I'm not sure that's right. I also moved the warning into module code and added a config param to disable it. Please let me know what you think.

File: compiler/optimizations/gpuTransforms.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (e-kayrakli) [reply compiler/optimizations/gpuTransforms.cpp:418]: Ah, I see. Would you be up for doing that before the release?

File: runtime/include/chpl-gpu.h [hunk 9]
```diff
@@ -201,6 +201,15 @@ GPU_CUB_WRAP(DECL_ONE_SORT, keys);
 
 #undef DECL_ONE_SORT
 
+#else // HAS_GPU_LOCALE
+
+// Provide a fallback for the chpl_assert_on_gpu function for non-GPU locales.
+// This works exactly the same as the standard one.
+
+static inline void chpl_assert_on_gpu(int32_t lineno, int32_t filenameIdx) {
```
Comment (e-kayrakli) [inline runtime/include/chpl-gpu.h:209]: Consider `ln` and `fn` for formal names to have consistency in the runtime. (I don't love those names at all, I love consistency)

File: compiler/optimizations/gpuTransforms.cpp (hunk 1 above)
Comment (e-kayrakli) [reply compiler/optimizations/gpuTransforms.cpp]: Looked at the current solution and I am ok with it. Thanks for making the change.

File: compiler/passes/convert-uast.cpp (hunk 3 above)
Comment (e-kayrakli) [reply compiler/passes/convert-uast.cpp:4373]: > Using an experimental annotation (which I plan to work on next after this PR) I can get chpldoc to use these function definitions (like chpl__assertOnGpuAttr) to generate documentation for attributes, which addresses

Nice.

Thanks for explanation. You are right on all counts.

File: compiler/optimizations/gpuTransforms.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (DanilaFe) [reply compiler/optimizations/gpuTransforms.cpp:418]: Yes, for sure.
